proposals/3392/dataset_partitioning.md

```diff
@@ -0,0 +1,331 @@
+                  "condition": {
+                     "type": "#/$defs/BaseFilterCondition"
+                  }
+               }
+            }
+         ]
+      }
+   },
+   "type": "object",
+   "properties": {
+      "filter": {
+         "$ref": "#/$defs/FilterInputDatasetFacet"
+      }
+   }
+}
+
+```
+
+
+
+An example filter facet for SQL `SELECT * FROM table WHERE col1 = 7 and col2 = 9` would be:
+
+```json
+{
+  "filter": {
+    "$id": "https://openlineage.io/spec/facets/1-0-0/BinaryFilterCondition.json",
+    "left": {
+      "field": "col1",
+      "value": "7",
+      "comparison": "EQUAL"
+    },
+    "right": {
+      "field": "col2",
+      "value": "9",
+      "comparison": "EQUAL"
+    },
+    "operator": "AND"
+  }
+}
+```
+    
+Please mind that this approach also allows providing a list of partitions, as well as extending it to add regex patterns on partitioning read.
+
+### Output partitions' dataset facet
+
+```json
+{
+   "$schema": "https://json-schema.org/draft/2020-12/schema",
+   "$defs": {
+      "BasePartitionOutputDatasetFacet": {
+         "allOf": [
+            {
+               "$ref": "#/$defs/OutputDatasetFacet"
+            }
+         ]
+      },
+      "SelectedPartitionsOutputDatasetFacet": {
+         "allOf": [
+            {
+               "$ref": "#/$defs/BasePartitionOutputDatasetFacet"
+            },
+            {
+               "type": "object",
+               "properties": {
+                  "operation": {
+                     "type": "string",
+                     "description": "Operation performed on a partition. Can be: 'ADD', 'UPDATE', 'DROP'."
+                  },
+                  "partitions": {
+                     "type": "array",
+                     "items": {
+                        "types": "object",
+                        "properties": {
+                           "location": {
+                              "type": "string",
+                              "description": "Optionally provided location of the partition specified"
+                           },
+                           "identifier": {
+                              "type": "string",
+                              "description": "Optionally provided identifier of the partition specified"
+                           }
+                        },
+                        "additionalProperties": true
+                     }
+                  }
+               }
+            }
+         ]
+      }
+   },
+   "type": "object",
+   "properties": {
+      "partitions": {
+         "$ref": "#/$defs/BasePartitionOutputDatasetFacet"
+      }
+   }
+}
+```
+
+An example of this would be a dataset partitioned by `business_date` and `country`. A job run affecting many countries for a single date would affect multiple partitions and this would be described with a facet:
+
+```json
+{
+   "$id": "https://openlineage.io/spec/facets/1-0-0/SelectedPartitionsOutputDatasetFacet.json",
+   "operation": "UPDATE",
```

Comment:
Theoretically, several operations could be performed in the same time. For example, writing dataframe with `df.write.mode("overwrite")` and `SparkSession.builder.config("spark.sql.sources.partitionOverwriteMode", "dynamic")` may:
* drop existing partitions, already presented in a filesystem and matched partition column values in the dataframe
* create new partitions, if partition column values were present in the dataframe only, but not in the filesystem

But if I understand it correctly, Spark currently doesn't report which partitions were dropped/created in this case. So the only way to capture exact partitions info is to parse commands manually, like AddPartitionCommand, AlterPartitionCommand and so on, which usually perform only one operation with specific list of partitions.

Reply:
I am not sure if one could support dynamic partitioning automatically. 
But it would definitely make sense to change `operation` property location and have it per partitions item.